### PR TITLE
[BREAKING CHANGE] Ignore added_tokens (both special and not) in the decoder

### PR DESCRIPTION
**Pull request by Narsil**
Causes issues with `ByteLevel` messing up some `AddedTokens` with some
utf-8 range used in the bytelevel mapping.

This commit tests the extend of the damage of ignoring the decoder for
those tokens.

### CODE REVIEW

File: tokenizers/src/tokenizer/mod.rs [hunk 1]
```diff
@@ -845,22 +847,35 @@ where
...
+                }
+                let text_chunk = if let Some(decoder) = &self.decoder {
+                    decoder.decode(chunks.clone())?
+                } else {
+                    chunks.join(" ")
+                };
+                result.push_str(&text_chunk);
+                if !result.is_empty() {
+                    result.push(' ');
+                }
```
Comment (ArthurZucker) [inline tokenizers/src/tokenizer/mod.rs:865]: why are we adding an extra space before the added token? IMO we should only do this if there is no decoder (default join on `" "`)

**Review by ArthurZucker — COMMENTED** [review]
LGTM, let's add a test for the decoder to make sure the behavior we are looking for is enable

File: tokenizers/src/tokenizer/mod.rs (hunk 1 above)
Comment (Narsil) [reply tokenizers/src/tokenizer/mod.rs:865]: https://github.com/huggingface/tokenizers/actions/runs/8819865853/job/24212114706#step:12:262

Comment (ArthurZucker) [reply tokenizers/src/tokenizer/mod.rs:865]: These test are done when there is no `decoder` no? 
I mean the previous behaviour does not add spaces between added token if there is a decoder

**Comment by thusinh1969** [comment]
Gents,
When can this fix be ready please ?

Thanks,
Steve

File: tokenizers/src/tokenizer/mod.rs (hunk 1 above)
Comment (Narsil) [reply tokenizers/src/tokenizer/mod.rs:865]: Yes it did (because the test wasn't using a decoder, so by default no decoder means space separated)

**Comment by Narsil** [comment]
More tests with actual failures 

```
FAILED tests/models/bartpho/test_tokenization_bartpho.py::BartphoTokenizerTest::test_clean_up_tokenization_spaces - assert "[CLS]this sh...' ll go.[SEP]" == "[CLS] thi...
FAILED tests/models/blenderbot_small/test_tokenization_blenderbot_small.py::BlenderbotSmallTokenizerTest::test_clean_up_tokenization_spaces - assert "[CLS]this sh...' ll go.[SEP]" == "[CLS] thi...
FAILED tests/models/idefics/test_processor_idefics.py::IdeficsProcessorTest::test_tokenizer_left_padding - AssertionError: '<unk[20 chars]><unk><unk><unk><unk...
FAILED tests/models/idefics/test_processor_idefics.py::IdeficsProcessorTest::test_tokenizer_padding - AssertionError: '<s>Describe this image.\nAssistant...
FAILED tests/models/luke/test_tokenization_luke.py::LukeTokenizerTest::test_clean_up_tokenization_spaces - assert "[CLS]this sh...' ll go.[SEP]" == "[CLS] thi...
FAILED tests/models/mluke/test_tokenization_mluke.py::MLukeTokenizerTest::test_clean_up_tokenization_spaces - assert "[CLS]this sh...' ll go.[SEP]" == "[CLS] thi...
FAILED tests/models/mpnet/test_tokenization_mpnet.py::MPNetTokenizerTest::test_clean_up_tokenization_spaces - assert "[CLS]this sh...' ll go.[SEP]" == "[CLS] thi...
FAILED tests/models/speech_to_text/test_tokenization_speech_to_text.py::SpeechToTextTokenizerTest::test_clean_up_tokenization_spaces - assert "[CLS]this sh...' ll go.[SEP]" == "[CLS] thi...
FAILED tests/models/speech_to_text_2/test_tokenization_speech_to_text_2.py::SpeechToTextTokenizerTest::test_clean_up_tokenization_spaces - assert "[CLS]this sh...' ll go.[SEP]" == "[CLS] thi...
FAILED tests/models/speecht5/test_tokenization_speecht5.py::SpeechT5TokenizerTest::test_clean_up_tokenization_spaces - assert "[CLS]this sh...' ll go.[SEP]" == "[CLS] thi...
FAILED tests/models/vits/test_tokenization_vits.py::VitsTokenizerTest::test_clean_up_tokenization_spaces - assert "[CLS]this sh...' ll go.[SEP]" == "[CLS] thi...
FAILED tests/models/wav2vec2/test_tokenization_wav2vec2.py::Wav2Vec2CTCTokenizerTest::test_clean_up_tokenization_spaces - assert "[CLS]this sh...' ll go.[SEP]" == "[CLS] thi...
FAILED tests/models/whisper/test_tokenization_whisper.py::WhisperTokenizerTest::test_clean_up_tokenization_spaces - assert "[CLS]this sh...' ll go.[SEP]" == "[CLS] thi...
```

File: tokenizers/src/tokenizer/mod.rs [hunk 2]
```diff
@@ -845,22 +847,35 @@ where
...
-
-        if let Some(decoder) = &self.decoder {
-            decoder.decode(tokens)
-        } else {
-            Ok(tokens.join(" "))
+        let mut result = String::with_capacity(ids.len());
+        let mut chunks = Vec::with_capacity(ids.len());
+        for id in ids {
+            if let Some(added_token) = self.added_vocabulary.simple_id_to_token(*id) {
+                if skip_special_tokens && self.added_vocabulary.is_special_token(&added_token) {
```
Comment (ArthurZucker) [inline tokenizers/src/tokenizer/mod.rs:854]:
```suggestion
                if skip_special_tokens && added_token.special) {
```
should be more foolproof?

File: tokenizers/src/tokenizer/mod.rs [hunk 3]
```diff
@@ -845,22 +847,35 @@ where
...
+                }
+                let text_chunk = if let Some(decoder) = &self.decoder {
+                    decoder.decode(chunks.clone())?
+                } else {
+                    chunks.join(" ")
+                };
+                result.push_str(&text_chunk);
+                if !result.is_empty() && self.decoder.is_none() {
+                    result.push(' ');
+                }
```
Comment (ArthurZucker) [inline tokenizers/src/tokenizer/mod.rs:865]: If there is no decoder, we are adding a space before the added token, but that is more backward compatible and more "logical" in this case. FIne for me!

**Review by ArthurZucker — APPROVED** [review]

File: tokenizers/src/tokenizer/mod.rs (hunk 2 above)
Comment (ArthurZucker) [reply tokenizers/src/tokenizer/mod.rs:854]: actually no worries

**Comment by ArthurZucker** [comment]
The failing test is basically the same one run on bert

**Review by ArthurZucker — COMMENTED** [review]
Actually this kinda breaks the way we decode, because we don't decode everything, meta space or anything that relies on the idx of the token is broken

**Comment by ArthurZucker** [comment]
I'll fix it

**Pull request merged**